### PR TITLE
RSA PSS signing

### PR DESCRIPTION
**Pull request by jschneider-bensch**
This PR addresses RSA-PSS signing in `tls13crypto`.

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Documentation
- [ ] Automation

## Motivation and Context
RSA-PSS is among the mandatory signature algorithms to implement for RFC 8446.

## Changes
I've extended `sign` in `tls13crypto` to handle the RSA-PSS case, and also adapted `certificate_verify` to build the correct message when sending out an RSA signature.

Some specific things, where I'm not sure how I did it is necessarily how it should be done:
- I'm giving a new argument `cert` to `sign`, so that `libcrux` can extract an RSA public key. This public key can then be used to build the actual signing key given the private exponent bytes in the `sk` argument. This meant changing the way `sign` is called in one location where we have the certificate around anyway. Alternatively we would have to get all the information for `libcrux` from `sk`, which probably means parsing a full ASN.1 private key from those bytes, i.e. extending the minimal parser in `tls13cert` to do that. The way I did it now seemed like an okay compromise to me.
- Previously, when creating the `certificate_verify` message for ECDSA certificates, it was clear how long the signature should be and this is actually checked. In the case of RSA signatures the length depends on the length of the key, which is not available at the point where the check should take place (I think).  Must this length be checked or is that more a sanity check for the steps before?
- 

## Checklist
- [ ] I have linked an issue to this PR
- [x] I have described the changes
- [x] I have read and understood the code of conduct, contribution guidelines, and contributor license agreement
- [x] I have added tests for all changes (in a separate PR, #70 )
- [x] I have tested that all tests pass locally and all checks pass
- [ ] I have updated documentation if necessary

Fixes #

### CODE REVIEW

File: src/tls13crypto.rs [hunk 1]
```diff
@@ -237,15 +240,48 @@ pub fn to_libcrux_sig_alg(a: &SignatureScheme) -> Result<signature::Algorithm, T
 pub fn sign(
     alg: &SignatureScheme,
     sk: &Bytes,
+    cert: &Bytes,
     input: &Bytes,
     ent: Bytes,
```
Comment (franziskuskiefer) [inline src/tls13crypto.rs:245]: Please add a comment here that this needs to be reworked.

File: src/tls13crypto.rs [hunk 2]
```diff
@@ -237,15 +240,48 @@ pub fn to_libcrux_sig_alg(a: &SignatureScheme) -> Result<signature::Algorithm, T
...
+
+            let sk = signature::rsa_pss::RsaPssPrivateKey::new(&pk, &sk.declassify())
+                .map_err(|_| CRYPTO_ERROR)?;
+
+            sk.sign(
+                signature::DigestAlgorithm::Sha256,
+                &salt,
+                &input.declassify(),
+            )
+            .map(|sig| signature::Signature::RsaPss(sig))
```
Comment (franziskuskiefer) [inline src/tls13crypto.rs:282]: What clippy says.

File: src/tls13formats.rs [hunk 3]
```diff
@@ -906,13 +906,19 @@ fn parse_ecdsa_signature(sig: Bytes) -> Result<Bytes, TLSError> {
...
+            if cv.len() != 64 {
+                return tlserr(parse_failed());
+            } else {
+                ecdsa_signature(cv)?
+            }
+        }
+    };
+
+    let sig = signature_algorithm(algs)?.concat(&lbytes2(&sv)?);
+    return Ok(handshake_message(HandshakeType::CertificateVerify, &sig)?);
```
Comment (franziskuskiefer) [inline src/tls13formats.rs:921]: What clippy says.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Generally lgtm. The design choices make sense to me. We might clean some of that up later but it certainly makes sense here.
Let's get #70 in first, then this will have tests on ci.

File: src/tls13crypto.rs (hunk 1 above)
Comment (jschneider-bensch) [reply src/tls13crypto.rs:245]: I've made an issue for it, #72.

**Comment by jschneider-bensch** [comment]
When #70 is in I should remove the `should_panic` from the RSA tests, too, before we merge this.

File: src/tls13crypto.rs [hunk 4]
```diff
@@ -240,7 +240,7 @@ pub fn to_libcrux_sig_alg(a: &SignatureScheme) -> Result<signature::Algorithm, T
 pub fn sign(
     alg: &SignatureScheme,
     sk: &Bytes,
-    cert: &Bytes,
+    cert: &Bytes, // TODO: `cert` added to allow reconstructing full signing key for RSA-PSS. Rework this. (cf. issue #72)
     input: &Bytes,
     ent: Bytes,
```
[the remark this answers is not in the document]
Comment (jschneider-bensch) [reply src/tls13crypto.rs]: That makes sense. :)
I've opened #73 for this.

**Review by franziskuskiefer — APPROVED** [review]
Thanks

**Pull request merged**